### PR TITLE
Allow PeaksWorkspace createPeakQSample to calculate goniometer for CW

### PR DESCRIPTION
**Pull request by rosswhitfield**
This involves moving the code implement in #21726 from [FindPeaksMD](http://docs.mantidproject.org/nightly/algorithms/FindPeaksMD-v1.html#calculate-goniometer-for-constant-wavelength) into new method on Goniometer. And added option onto peaks workspace where by the property `wavelength` exist it will calculate the goniometer.

**To test:**
```
LoadMD(Filename='/SNS/users/rwp/IPTS-20367_Q_sample_slice.nxs', LoadHistory=False, OutputWorkspace='q')

LoadEventNexus(Filename='/HFIR/HB2C/IPTS-20367/nexus/HB2C_15954.nxs.h5', OutputWorkspace='ws',MetaDataOnly=True)
peaks = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0)
peaks.run().addProperty('wavelength', 1.488, True)
# or
peaks.run().addProperty('energy', 36.94, True)
```
From here open up slice viewer, overlay peaks workspace, then add peaks using the `Add HKL Peaks` button. The peaks added should now have the correct wavelength, d-spacing, Q lab, Q sample and detector information.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataObjects/src/PeaksWorkspace.cpp [hunk 1]
```diff
@@ -278,7 +279,26 @@ PeaksWorkspace::createPeak(const Kernel::V3D &position,
...
+
+  Geometry::Goniometer goniometer;
+
+  LogManager_const_sptr props = getLogs();
+  if (props->hasProperty("wavelength") || props->hasProperty("energy")) {
+    // Assume constant wavelenth
+    // Calculate Q lab from Q sample and wavelength
+    double wavelength;
+    if (props->hasProperty("energy")) {
+      wavelength = Kernel::UnitConversion::run(
```
Comment (AndreiSavici) [inline Framework/DataObjects/src/PeaksWorkspace.cpp]: Convert directly to momentum?

File: Framework/DataObjects/src/PeaksWorkspace.cpp [hunk 2]
```diff
@@ -278,7 +279,26 @@ PeaksWorkspace::createPeak(const Kernel::V3D &position,
...
+          "Energy", "Wavelength",
+          props->getPropertyValueAsType<double>("energy"), 0, 0, 0,
+          Kernel::DeltaEMode::Elastic, 0);
+    } else {
+      wavelength = props->getPropertyValueAsType<double>("wavelength");
+    }
+    goniometer.calcFromQSampleAndWavelength(position, wavelength);
```
Comment (AndreiSavici) [inline Framework/DataObjects/src/PeaksWorkspace.cpp]: Convert to momentum?

File: Framework/Geometry/inc/MantidGeometry/Instrument/Goniometer.h [hunk 3]
```diff
@@ -91,6 +91,10 @@ class MANTID_GEOMETRY_DLL Goniometer {
   // Set rotation angle for an axis in the units the angle is set (default --
   // degrees)
   void setRotationAngle(size_t axisnumber, double value);
+  // Calculate goniometer for rotation around y-axis for constant wavelength
+  // from Q Sample
+  void calcFromQSampleAndWavelength(const Mantid::Kernel::V3D &Q,
```
Comment (AndreiSavici) [inline Framework/Geometry/inc/MantidGeometry/Instrument/Goniometer.h]: If converted to momentum, this needs to be renamed

File: docs/source/algorithms/FindPeaksMD-v1.rst [hunk 4]
```diff
@@ -83,8 +83,12 @@ First calculate the :math:`\textbf{Q}_{lab}` using
 
 .. math:: \theta = \cos^{-1}(1-\frac{|\textbf{Q}_{sample}|^2}{2k^2})
 
+where :math:`\theta` is from 0 to :math:`\pi`
+
 .. math:: \phi = \sin^{-1}(-\textbf{Q}_{sample}^y \sin(\theta)/k)
 
+where :math:`\phi` is from :math:`-\pi/2` to :math:`\pi/2`
+
```
Comment (AndreiSavici) [inline docs/source/algorithms/FindPeaksMD-v1.rst]: Need to explicitly tell that this means that you will get a detector position on the left hand side of the beam, even if you don't have a detector there.

**Review by AndreiSavici — CHANGES_REQUESTED** [review]
You can ignore the first comments about momentum, if you want to. Documentation change is required

File: Framework/Geometry/inc/MantidGeometry/Instrument/Goniometer.h (hunk 3 above)
Comment (rosswhitfield) [reply Framework/Geometry/inc/MantidGeometry/Instrument/Goniometer.h]: Using wavelength is better for most of my use cases.

**Review by AndreiSavici — APPROVED** [review]

**Review by AndreiSavici — CHANGES_REQUESTED** [review]
The example fails if the crystallography convention is selected

**Review by AndreiSavici — APPROVED** [review]

File: Framework/Geometry/src/Instrument/Goniometer.cpp [hunk 5]
```diff
@@ -178,6 +179,42 @@ void Goniometer::setRotationAngle(size_t axisnumber, double value) {
...
+            wv * (1 - cos(theta)));
+
+  // Solve to find rotation matrix, assuming only rotation around y-axis
+  // A * X = B
+  Matrix<double> A({Q[0], Q[2], Q[2], -Q[0]}, 2, 2);
+  A.Invert();
+  std::vector<double> B{Q_lab[0], Q_lab[2]};
+  std::vector<double> X = A * B;
+  double rot = atan2(X[1], X[0]);
+  g_log.information() << "Found goniometer rotation to be " << rot * 180 / M_PI
```
Comment (martyngigg) [inline Framework/Geometry/src/Instrument/Goniometer.cpp]: Should we remove the logging at this level? `Goniometer` feels quite low down to be logging at information at least.

Comment (martyngigg) [reply Framework/Geometry/src/Instrument/Goniometer.cpp]: :+1:

**Pull request merged**